typing/types.ml

```diff
@@ -497,31 +501,64 @@ let commu_var () = Cvar {commu=Cunknown}
 
 let rec repr_link (t : type_expr) d : type_expr -> type_expr =
  function
-   {desc = Tlink t' as d'} ->
+   {desc = Tlink t' | Texpand (t', _, _) as d'} ->
+     let d' = match d with
+     | Texpand (_, path, args) -> Texpand (t', path, args)
+           (* keep the original name and args if they exist *)
+           (* NB: Texpand is allocated at each recursion;
+              could be optimized *)
+     | _ -> d'
+     in
      repr_link t d' t'
  | {desc = Tfield (_, k, _, t') as d'}
    when field_kind_internal_repr k = FKabsent ->
      repr_link t d' t'
+     (* In the case of Tfield, d does not contain an abbreviation,
+        thus we do not need a case analysis like above. *)
  | t' ->
      log_change (Ccompress (t, t.desc, d));
      t.desc <- d;
      t'
 
-let repr_link1 t = function
-   {desc = Tlink t' as d'} ->
+let repr_link1 t d = function
+   {desc = Tlink t' | Texpand (t', _, _) as d'} ->
+     let d' = match d with
+     | Texpand (_, path, args) -> Texpand (t', path, args)
+     | _ -> d'
+     in
      repr_link t d' t'
  | {desc = Tfield (_, k, _, t') as d'}
    when field_kind_internal_repr k = FKabsent ->
      repr_link t d' t'
  | t' -> t'
 
+(* A non-normalized type may contain an arbitrarily long chain of
+   [Tlink] or [Texpand] prefixing the actual node (neither
+   [Tlink] nor [Texpand]).
+   [repr t] returns the actual node [t'], and also normalizes [t]
+   so that it is either [t'] itself (if there was no prefix) or
+   either [Tlink t'] or [Texpand (t', path, args)] where [path] and [args]
+   are taken from the first [Texpand] in the chain.
+   For instance, with "t1 -> t2" for [t1.desc = Tlink t2] and
+   "t1 -(path,args)-> t2" for [t1.desc = Texpand (t2, path, args)], we have
+   Before:
+      t -> t1 -(path1,args2)-> t2 -> t3 -(path2,args2)-> t'
+   After:
+      t -(path1,args1)-> t'
+   Before:
+      t -> t1 -> t2 -> t'
+   After:
+      t -> t'
+ *)
+
 let repr t =
-  match t.desc with
-   Tlink t' ->
-     repr_link1 t t'
- | Tfield (_, k, _, t') when field_kind_internal_repr k = FKabsent ->
-     repr_link1 t t'
- | _ -> t
+  let d = t.desc in
+  match d with
+    Tlink t' | Texpand (t', _, _) ->
+      repr_link1 t d t'
+  | Tfield (_, k, _, t') when field_kind_internal_repr k = FKabsent ->
+      repr_link1 t d t'
+  | _ -> t
 
 (* scope_field and marks *)
```

Comment:
Maybe we can introduce the following function to remove the usages of `ignore` below? It also documents the intention to normalize the `type_expr`

```suggestion
  | _ -> t

let normalize t =
  (ignore : type_expr -> unit) (repr t)

(* scope_field and marks *)
```